feat: add multi-instance support

elias-pap opened the pull request:
closes #255

edno commented:
Thank you @elias-pap for this new feature!!! That's awesome.

I will deep dive into the code asap.

In the meanwhile, could you add some tests that cover the multi instance case?

elias-pap commented:
> Thank you @elias-pap for this new feature!!! That's awesome.
> 
> I will deep dive into the code asap.
> 
> In the meanwhile, could you add some tests that cover the multi instance case?

Thanks for this useful plugin @edno !

edno commented on src/index.js:
```diff
@@ -2,7 +2,8 @@
 const path = require("path");
 const os = require("os");
-const GroupInfo = require("./lib/group-info");
+const { pluginConfigs } = require("./utils/helpers/other.js");
```
`pluginConfigs[]` should not be exposed, see comment regarding `pluginConfigs.push({ ...DEFAULT_OPTIONS, ...opts });`

elias-pap replied:
Addressed in 9b550e2.

edno commented on src/index.js:
```diff
@@ -17,65 +18,38 @@ const DEFAULT_OPTIONS = {
-              config.groupByDirective,
-            prettify: options.pretty || config.pretty,
-          });
+          for (const config of pluginConfigs) {
+            await generateDocFromSchema(
+              mergeConfigWithCLIOptions(config, options),
+            );
+          }
```
Let's create a `run()` function to simplify the code:

```js
\\\....
        .action(async (options) => {
          await run(pluginConfigs, options);
        });
    },
  };
};

async function run(pluginConfigs, cliOptions) {
  for (const config of pluginConfigs) {
    await generateDocFromSchema(buildConfig(config, cliOptions));
  }
}
```

elias-pap replied:
Not needed since we chose the "one command instance per config" approach.

edno commented on src/index.js:
```diff
@@ -17,65 +18,38 @@ const DEFAULT_OPTIONS = {
 module.exports = function pluginGraphQLDocGenerator(context, opts) {
-  // Merge defaults with user-defined options.
-  const config = { ...DEFAULT_OPTIONS, ...opts };
+  // Merge defaults with user-defined options in config file.
+  pluginConfigs.push({ ...DEFAULT_OPTIONS, ...opts });
```
`pluginConfigs`should be immutable. This can be done by using a builder method. 
At the same time, we can also move `DEFAULT_OPTIONS` and the merge logic with it to `config.js`.

```js
const path = require("path");
const os = require("os");

const DEFAULT_OPTIONS = {
  schema: "./schema.graphl",
  rootPath: "./docs",
  baseURL: "schema",
  linkRoot: "/",
  homepage: path.join(__dirname, "../assets/", "generated.md"),
  diffMethod: "SCHEMA-DIFF",
  tmpDir: path.join(os.tmpdir(), "@edno/docusaurus2-graphql-doc-generator"),
  loaders: {},
  pretty: false,
};

const pluginConfigs = [];

function addConfig(options) {
  pluginConfigs.push({ ...DEFAULT_OPTIONS, ...options });
  return pluginConfigs;
}
```

elias-pap replied:
Cool. Addressed in 9b550e2 (there is no array anymore).

edno commented on tests/e2e/docusaurus2-graphql-doc-generator-multi-instance-1.config.js:
```diff
@@ -0,0 +1,10 @@
+  rootPath: "./docs-tweet-A",
+  baseURL: "schema",
+  linkRoot: "/",
+  loaders: {
+    UrlLoader: "@graphql-tools/url-loader",
+  },
```
not required
```suggestion
```

elias-pap replied:
Good catch. Addressed in 9b550e2.

edno commented on tests/e2e/docusaurus2-graphql-doc-generator-multi-instance-1.config.js:
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+  id: "config-A",
+  schema: "data/tweet-A.graphql",
+  rootPath: "./docs-tweet-A",
```
I'd rather limit the number of local schemas to maintain
```suggestion
  id: "tweet",
  schema: "data/tweet.graphql",
  rootPath: "./doc-tweet",
```

elias-pap replied:
Sure. Addressed in 9b550e2.

edno commented on tests/e2e/docusaurus2-graphql-doc-generator-multi-instance-2.config.js:
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+  id: "config-B",
+  schema: "data/tweet-B.graphql",
+  rootPath: "./docs-tweet-B",
```
I'd rather limit the number of local schemas to maintain
```suggestion
  id: "swapi",
  schema: "data/swapi.graphql",
  rootPath: "./doc-swapi",
```

elias-pap replied:
Addressed in 9b550e2.

edno commented on tests/e2e/docusaurus2-graphql-doc-generator-multi-instance-2.config.js:
```diff
@@ -0,0 +1,10 @@
+  rootPath: "./docs-tweet-B",
+  baseURL: "schema",
+  linkRoot: "/",
+  loaders: {
+    UrlLoader: "@graphql-tools/url-loader",
+  },
```
not required
```suggestion
```

elias-pap replied:
Addressed in 9b550e2.

edno commented on tests/e2e/docusaurus2-graphql-doc-generator-multi-instance.config.js:
```diff
@@ -0,0 +1,22 @@
+module.exports = [
+  {
+    id: "config-A",
+    schema: "data/tweet-A.graphql",
+    rootPath: "./docs-tweet-A",
+    baseURL: "schema",
+    linkRoot: "/",
+    loaders: {
+      UrlLoader: "@graphql-tools/url-loader",
+    },
+  },
+  {
+    id: "config-B",
+    schema: "data/tweet-B.graphql",
+    rootPath: "./docs-tweet-B",
+    baseURL: "schema",
+    linkRoot: "/",
+    loaders: {
+      UrlLoader: "@graphql-tools/url-loader",
+    },
+  },
```
Same as above
```suggestion
  {
    id: "tweet",
    schema: "data/tweet.graphql",
    rootPath: "./docs-tweet",
    baseURL: "schema",
    linkRoot: "/",
  },
  {
    id: "swapi",
    schema: "data/swapi.graphql",
    rootPath: "./docs-swapi",
    baseURL: "schema",
    linkRoot: "/",
  },
```

This can also be simplified to:
```js
[
  require('./docusaurus2-graphql-doc-generator-multi-instance-1.config.js'),
  require('./docusaurus2-graphql-doc-generator-multi-instance-2.config.js'),
]
```

elias-pap replied:
Nice refactoring. Addressed in 9b550e2.

edno commented on Earthfile:
```diff
@@ -79,11 +79,18 @@ smoke-test:
   COPY ./tests/e2e/jest.config.js ./jest.config.js
   ENV NODE_ENV=ci
   RUN node --expose-gc /usr/local/bin/jest --logHeapUsage --runInBand
+  COPY ./tests/e2e/docusaurus2-graphql-doc-generator-multi-instance.config.js ./docusaurus2-graphql-doc-generator-multi-instance.config.js
+  COPY ./tests/e2e/docusaurus2-graphql-doc-generator-multi-instance-1.config.js ./docusaurus2-graphql-doc-generator-multi-instance-1.config.js
+  COPY ./tests/e2e/docusaurus2-graphql-doc-generator-multi-instance-2.config.js ./docusaurus2-graphql-doc-generator-multi-instance-2.config.js
```
Files `./tests/e2e/*.config.js` should be located under `tests/__data__`.

elias-pap replied:
So should we move the existing ones `docusaurus2-graphql-doc-generator.config.js` and `docusaurus2-graphql-doc-generator-groups.config.js` as well ? My understanding was that `__data__` folder contains the raw schemas and some markdown files, while configuration is considered to be code.

edno replied:
Good catch! `__data__` should contain everything that is not a test.

elias-pap replied:
Alright, addressed in 0d26176.

edno commented on Earthfile:
```diff
@@ -79,11 +79,18 @@ smoke-test:
+  COPY ./tests/e2e/docusaurus2-graphql-doc-generator-multi-instance.config.js ./docusaurus2-graphql-doc-generator-multi-instance.config.js
+  COPY ./tests/e2e/docusaurus2-graphql-doc-generator-multi-instance-1.config.js ./docusaurus2-graphql-doc-generator-multi-instance-1.config.js
+  COPY ./tests/e2e/docusaurus2-graphql-doc-generator-multi-instance-2.config.js ./docusaurus2-graphql-doc-generator-multi-instance-2.config.js
+  COPY ./scripts/config-plugin-multi-instance.js ./config-plugin-multi-instance.js
+  COPY ./tests/e2e/specs/cli-multi-instance.spec.js ./__tests__/e2e/specs/cli.spec.js
+  RUN node config-plugin-multi-instance.js
+  RUN node --expose-gc /usr/local/bin/jest --logHeapUsage --runInBand
```
I do not understand why both `cli-multi-instance.spec.js` and `cli.spec.js` cannot be run at once.

elias-pap replied:
2 reasons:
1. `config-plugin.js` generates a `docusaurus.config.js` file, and `config-plugin-multi-instance.js` generates a different `docusaurus.config.js`. Single instance tests should run with the former and multi instance with the latter.
2. The code checks if the schema has changed to decide whether to generate the docs. This might make the tests flaky if they run in parallel because they take as input the same schema files. For example, if the multi-instance finishes first then the "should return 0 with generated message when completed as first run" test might fail (since this is no longer the first run).

edno replied:
Since the multi-instance is now aligned with Docusaurus, ie one command per instance. I suggest merging both tests. The aim being to check that a config without `id` will run correctly, and a second config with`id` will not conflict with the default config. I hope this make sense.

elias-pap replied:
Good point. Addressed in 0d26176.

edno reviewed (changes requested):
I did a review. A few comments, but nothing that requires major changes.

Great job!!

edno commented:
I am wondering if this feature should be closer to the current Docusaurus multi-instance approach, see https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-content-docs/src/index.ts#L71

elias-pap commented:
> I am wondering if this feature should be closer to the current Docusaurus multi-instance approach, see https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-content-docs/src/index.ts#L71
> 
> This would mean one command per instance (id), instead of one command for all. That would give a more granular control over the generation to users, and users can still merge both command into a single one with a script.

That's right, this is how I had implemented it initially and the change required is far more simpler:
```
const command = opts.id === "default" ? "graphql-to-doc" : `graphql-to-doc:${opts.id}`;
.
.
.
return {
    name: "docusaurus-graphql-doc-generator",

    extendCli(cli) {
      cli
        .command("graphql-to-doc")  // replace with .command(command) here
        .option("-s, --schema <schema>", "Schema location")
```

I changed my mind later on because this mutates the name of the command, and this seemed a bit weird. Docusaurus handles that detail internally for their `docs` plugin, so the user does not have to deal with many commands.

But yeah, we can do this too ! Whatever you think is best.

edno commented:
> I changed my mind later on because this mutates the name of the command, and this seemed a bit weird. Docusaurus handles that detail internally for their docs plugin, so the user does not have to deal with many commands.

Docusaurus also changes the command name (and description): https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-content-docs/src/index.ts#L89

Both approaches are valid. I would like to stay as close as possible to Docusausus behaviour for the v1 (the v2 will one day allow breaking the dependency to Docusaurus).

Ideally, I'd love to see the default command running all. But that will be a nightmare to solve this. And, to stick to KISS principle (I should add this to the contrib guidelines), I think we should go with Docusaurus command per `id`.

```js
const isDefaultId = opts.id === "default";

const command = isDefaultId ? "graphql-to-doc" : `graphql-to-doc:${opts.id}`;
const description = `Generate GraphQL Schema Documentation for ${opts.schema}`; // trick for helping users

return {
    name: "docusaurus-graphql-doc-generator",

    extendCli(cli) {
      cli
        .command(command)
        .description(description)   
```

And, we can keep the `config.js` logic for merging config with cli option. I like that approach, it makes the code cleaner.

elias-pap commented on src/index.js:
```diff
@@ -2,7 +2,8 @@
 const os = require("os");
-const GroupInfo = require("./lib/group-info");
+const { pluginConfigs } = require("./utils/helpers/other.js");
+const { mergeConfigWithCLIOptions } = require("./utils/helpers/other.js");
```
Addressed in 9b550e2.

elias-pap commented on tests/__data__/tweet-A.graphql:
```diff
@@ -0,0 +1,82 @@
+
+type Subscription {
+    Notifications(limit: Int): [Notification]!
+}
```
Addressed in 9b550e2 (deleted file).

elias-pap commented on tests/__data__/tweet-B.graphql:
```diff
@@ -0,0 +1,82 @@
+
+type Subscription {
+    Notifications(limit: Int): [Notification]!
+}
```
Addressed in 9b550e2 (deleted file).

elias-pap commented on src/index.js:
```diff
@@ -1,81 +1,40 @@
+  const isDefaultId = opts.id === "default";
+
+  const command = isDefaultId ? "graphql-to-doc" : `graphql-to-doc:${opts.id}`;
+  const description = isDefaultId
+    ? "Generate GraphQL Schema Documentation"
+    : `Generate GraphQL Schema Documentation for configuration with id ${opts.id}`;
```
Cannot use `opts.schema` in the description because the schema might be overriden from the CLI `--schema` option later.

edno replied:
Fine with it 👍

edno commented:
Released: https://github.com/edno/graphql-markdown/releases/tag/1.9.0 🎉
